docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,325 @@
+
+In particular, we are interested in warpgroup specialization on Hopper+ GPUs, where we have a separate warpgroup issuing TMAs (GMEM/SMEM copies) from the warpgroups performing arithmetic, since indexing calculations and issuing TMAs can take up a significant amount of time and potentially leave the TensorCore idle. The figure below depicts a standard, non-specialized kernel on the left where TMAs (async copies) and matrix multiplication are written to a single instruction stream, and a warp-specialized version on the right where communication and arithmetic are handled on separate warpgroups. A *consumed barrier* is used to synchronize between the specialized warpgroups that signals to the memory warpgroup when it is safe to begin the next TMA.
+
+
+<!-- #endregion -->
+
+<!-- #region id="n-y90IC7v7vL" -->
+
+![warp_specialization](../../_static/pallas/gpu/warp_specialization.svg)
+
+<!-- #endregion -->
+
+<!-- #region id="ZH0Pui5kFSdD" -->
+Warp specialization can be enabled in Pallas by using the `plgpu.emit_pipeline_warp_specialized` helper. This pipeline helper handles all of the logic in the memory warpgroup, and the user only needs to specify the work done in the compute warpgroups. It shares the a similar API as the standard `emit_pipeline`, and currently supports the following arguments:
+
+```python
+plgpu.emit_pipeline_warp_specialized(
+  body: Callable,
+  *
+  grid: tuple[int, ...],
+  in_specs: Sequence[pallas_core.BlockSpec] = (),
+  out_specs: Sequence[pallas_core.BlockSpec] = (),
+  max_concurrent_steps: int,
+  carry_coroutine: Callable
+  num_compute_wgs: int,
+  memory_registers: int
+  wg_axis: str,
+  memory_thread_idx: int | None = None,
+)
+```
+
+There are a few arguments specific to this pipeline emitter, which are:
+- `num_compute_wgs` specifies how many compute warpgroups to use. The pipeline emitter always uses a single memory warpgroup, so in `plgpu.kernel` you should specify `num_threads=num_compute_wgs+1`.
+- `memory_registers` controls how many registers to allocate to the memory warpgroup. The remaining registers are partitioned evenly among the compute warpgroups.
+- `wg_axis` the name of the warpgroup axis (as specified by the `thead_name` argument of `plgpu.kernel`).
+- `memory_thread_idx` specifies which warpgroup to designate as the memory warpgroup. Defaults to the last warpgroup.
+- `carry_coroutine` is a Python coroutine that enables you to carry values in registers through the pipeline. The coroutine runs only on the compute warpgroups and defines the initialization of the carry and consumption of the final carry. All compute warpgroup specific arrays should be instantiated here so the memory warpgroup does not materialize them in registers -- otherwise, you may experience slowdowns due to register spills.
```

Comment:
You never explain what "the carry" really is. Please explain that it's the state that is carried through the pipeline steps, so this argument lets you specify a prologue/epilogue of the pipeline that will only run int he compute warpgroups